packages/jupyter-ai-magics/jupyter_ai_magics/magics.py

```diff
@@ -500,6 +469,26 @@ def run_ai_cell(self, args: CellArgs, prompt: str):
             self.transcript_openai = []
             return
 
+        # Apply a prompt template.
+        prompt = Provider.prompt_template(args.format, local_model_id).format(
+            prompt=prompt
+        )
+
+        # interpolate user namespace into prompt
+        ip = get_ipython()
+        prompt = prompt.format_map(FormatDict(ip.user_ns))
+
+        # Determine provider and local model IDs
+        # If this is a custom chain, send the message to the custom chain.
+        if args.model_id in self.custom_model_registry and isinstance(
+            self.custom_model_registry[args.model_id], LLMChain
+        ):
+            return self.display_output(
+                self.custom_model_registry[args.model_id].run(prompt),
+                args.format,
+                {"jupyter_ai": {"custom_chain_id": args.model_id}},
+            )
```

Comment:
I noticed that this block skips the auth validation block immediately below; honestly auth validation should be a class method on `BaseProvider` that's available and overridable by all providers, and this code should just call `Provider.validate_authn()`. This doesn't need to be fixed in this PR, as it is a pre-existing issue. I've created a separate issue to track this: https://github.com/jupyterlab/jupyter-ai/issues/311